### PR TITLE
chore(gtm): add Google Tag Manager custom template

### PR DESCRIPTION
**Pull request by francoischalifour**
_This adds the [Google Tag Manager](https://developers.google.com/tag-manager/) (GTM) [Custom Template](https://developers.google.com/tag-manager/templates/) for Search Insights._

Custom templates let you create a GTM interface to trigger events based on user interactions. This implementation is based on previous work by @thyngster.

## Workflow

The source code is in `src/`. All the separate parts are contained in their own files. This has been done to better leverage tooling (Prettier, IDE file extension support, etc.) and to ease reviews.

The `build` script generates the final custom template in `generated/search-insights.tpl` (note that the filename is not versioned, any quirks?).

## Library versioning

The [jsDelivr Search Insights 1.2.2 endpoint](https://cdn.jsdelivr.net/npm/search-insights@1.2.2) is the default endpoint targeted by the template. It can be overridden in the template configuration (see README > FAQ).

## Template versioning

The template is versioned to gather usage metrics with `addAlgoliaAgent`.

The template versioning is quite primitive for now as there is no standards yet (see README > Contributing).

I went for a semver versioning hardcoded inside the template in the `TEMPLATE_VERSION` variable (we can introduce more tooling later on). Perhaps that semver doesn't make sense for this tool because we're not supposed to introduce any breaking changes anyway?

## JavaScript sandbox

The JavaScript code in `template.js` is executed in a sandbox environment that doesn't support all ECMAScript features (e.g. string literals, spread operator, etc.).

The `require` statements do not import external libraries like we're used to, but rather some helpers provided by GTM.

## How to review

- You can either check the diff in the PR, or check [this PR branch](https://github.com/algolia/search-insights.js/tree/chore/gtm-template/gtm) since all files are new
- Log in to [Google Tag Manager](https://tagmanager.google.com/) and follow the README

### CODE REVIEW

File: gtm/README.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+</div>
+
+---
+
+## Usage
+
+### 1. Import the template
+
+- Download **[`search-insights.tpl`](generated/search-insights.tpl)**
+- Import the file in [Google Tag Manager](https://tagmanager.google.com) (GTM)
```
Comment (Haroenv) [inline gtm/README.md]:
```suggestion
- Import the file in [Google Tag Manager](https://tagmanager.google.com)
```

File: gtm/README.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
+# Search Insights for Google Tag Manager
```
Comment (Haroenv) [inline gtm/README.md]:
```suggestion
# Search Insights for Google Tag Manager (GTM)
```

File: gtm/README.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+- Import the file in [Google Tag Manager](https://tagmanager.google.com) (GTM)
+  - Go to "Templates"
+  - Click on "New"
+  - Select "Import" in the menu
+
+### 2. Create variables
+
+For events to be sent with the correct information (`userToken`, `objectIDs`, `queryID`, etc.), you need to add variables in the "User-Defined Variables" section.
+
+[Learn more about GTM variables →](https://www.simoahava.com/analytics/variable-guide-google-tag-manager/)
```
Comment (Haroenv) [inline gtm/README.md]: is there an official guide on this?

File: gtm/README.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+The user token can be given to GTM via a [Data Layer](https://developers.google.com/tag-manager/devguide#datalayer).
+
+```js
+window.dataLayer = window.dataLayer || [];
+window.dataLayer.push({
+  userToken: 'USER_1',
+});
+```
+
+### How to get the object IDs?
```
Comment (Haroenv) [inline gtm/README.md]: Do we need a similar one for `position`?

File: gtm/README.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+### How to change the Search Insights endpoint?
+
+By default, the Search Insights endpoint targets [jsDelivr CDN](https://www.jsdelivr.com/). If you want to use a specific version of the Search Insights library or to host your own version of the library, you can change the endpoint:
+
+1. In the "Init" event configuration, go to "Advanced Algolia Settings" and update "Search Insights Endpoint"
+1. In the "Permissions" template, add the same link in the "[Inject Scripts](https://www.simoahava.com/analytics/custom-templates-guide-for-google-tag-manager/#injects-scripts)" section
+
+## Contributing
+
+Before starting working on the project, make sure to **disable any ad blockers**.
```
Comment (Haroenv) [inline gtm/README.md]: 😂

File: README.md [hunk 6]
```diff
@@ -7,6 +7,8 @@ Search Insights lets you report click, conversion and view metrics using the [Al
 
 ## Getting started
 
+> Are you using Google Tag Manager in your app? We provide a [custom template](gtm) to ease the integration.
```
Comment (tkrugg) [inline README.md]: did you mean to add the link here?
is `gtm` pointing to anything?

Comment (Haroenv) [reply README.md]: gtm is the folder

File: gtm/README.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+### 1. Import the template
+
+- Download **[`search-insights.tpl`](generated/search-insights.tpl)**
+- Import the file in [Google Tag Manager](https://tagmanager.google.com) (GTM)
+  - Go to "Templates"
+  - Click on "New"
+  - Select "Import" in the menu
+
+### 2. Create variables
```
Comment (tkrugg) [inline gtm/README.md]: Can we help out the user more here, by explaining which variables need to be added, and where to find them?

File: gtm/README.md (hunk 4 above)
Comment (tkrugg) [reply gtm/README.md]: I agree it's might be worth documenting positions as well.
Actually since the steps are exactly the same for `objectID`, `position` and `queryID` maybe we can group them together and give the user a single snippet to copy-paste?

File: gtm/README.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+> `__queryID` is available from:
+>
+> - InstantSearch.js 3.4.0
+> - React InstantSearch 5.5.0
+> - Vue InstantSearch 2.1.0
+> - Angular InstantSearch 3.0.0
+
+### How to change the Search Insights endpoint?
+
+By default, the Search Insights endpoint targets [jsDelivr CDN](https://www.jsdelivr.com/). If you want to use a specific version of the Search Insights library or to host your own version of the library, you can change the endpoint:
```
Comment (tkrugg) [inline gtm/README.md]: I think using the word "endpoint" when meaning "source url hosting the library" can be a little misleading. It can be confused with the insights API endpoint: https://www.algolia.com/doc/rest-api/insights/#overview

File: gtm/README.md (hunk 5 above)
Comment (tkrugg) [inline gtm/README.md]: Is it not 

```suggestion
Before starting to work on the project, make sure to **disable any ad blockers**.
```

?

File: gtm/README.md (hunk 8 above)
Comment (tkrugg) [reply gtm/README.md]: Maybe we can reuse the variable from the snippet? `ALGOLIA_INSIGHTS_SRC`
https://www.algolia.com/doc/api-reference/widgets/hits/js/#install-the-insights-api-client-for-javascript

File: gtm/README.md (hunk 3 above)
Comment (francoischalifour) [reply gtm/README.md]: Simo Ahava is the reference for this for now, he basically writes the documentation that Google is missing...

File: gtm/README.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+1. In the "Init" event configuration, go to "Advanced Algolia Settings" and update "Search Insights Endpoint"
+1. In the "Permissions" template, add the same link in the "[Inject Scripts](https://www.simoahava.com/analytics/custom-templates-guide-for-google-tag-manager/#injects-scripts)" section
+
+## Contributing
+
+Before starting working on the project, make sure to **disable any ad blockers**.
+
+### Folder structure
+
+The source for the GTM template is in the [`src/`](src) folder and gets generated in the [`generated/`](generated) folder.
```
Comment (tkrugg) [inline gtm/README.md]: Would it be more obvious for someone stepping in this repo if we called the folder `generated` → `google-tag-manager`?

File: gtm/README.md (hunk 7 above)
Comment (francoischalifour) [reply gtm/README.md]: Sure, we can guide them more. For now I didn't want to give too many indications and "good practices" because I'm myself not entirely sure what to recommend, hence the link to Simo Ahava's blog. Let's wait a bit what patterns emerge before recommending too many things (custom templates are still quite new).

File: gtm/README.md (hunk 4 above)
Comment (francoischalifour) [reply gtm/README.md]: That's right, I'll add a part on `__position`.

File: gtm/README.md (hunk 8 above)
Comment (francoischalifour) [reply gtm/README.md]: I see what you mean – I'll rephrase it.

File: gtm/README.md (hunk 5 above)
Comment (francoischalifour) [reply gtm/README.md]: I think there's a nuance between the two. "Starting to do" something rather refers to "starting a new habit" while "starting doing" is more in "the now".

Reference: https://english.stackexchange.com/questions/204264/start-to-inf-vs-start-v-ing

Anyway, I can rephrase it to "Before working on the project" to remove the confusion.

Comment (Haroenv) [reply gtm/README.md]: Your last suggestion fixes the error!

File: gtm/src/template.js [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+      region: data.region,
+      cookieDuration: data.cookieDuration,
+    };
+
+    logger(data.eventType, initOptions);
+
+    aa('init', initOptions);
+
+    if (data.setUserToken) {
+      aa('setUserToken', data.setUserToken);
```
Comment (tkrugg) [inline gtm/src/template.js]: I'm not sure how this is used. It is supposed to be just text if I look at the spec of the template but here it's injected like it was a function.
Do you have a codesandbox demonstrating the usage?

File: gtm/README.md (hunk 7 above)
Comment (tkrugg) [reply gtm/README.md]: I just meant listing the variables they need to create and tell them were to find them, like you did for queryID for example. Does this need to wait for patterns to emerge?

File: gtm/src/info.json [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+    "displayName": "Algolia",
```
Comment (tkrugg) [inline gtm/src/info.json]: Algolia Insights may be a more accurate name?

File: gtm/src/template-parameters.json [hunk 12]
```diff
@@ -0,0 +1,404 @@
...
+        "value": "viewedObjectIDs"
+      },
+      {
+        "displayValue": "Viewed Filters",
+        "value": "viewedFilters"
+      }
+    ],
+    "displayName": "Event Type",
+    "simpleValueType": true,
+    "name": "eventType",
```
Comment (tkrugg) [inline gtm/src/template-parameters.json]: it would be more accurate to call this method, since it can also be `'init'`, which is not an event for insights.
Event type in the library is an enum: https://github.com/algolia/search-insights.js/blob/927e74e9b2dcdbb3ebe35fae7d30422987d8a33c/lib/_sendEvent.ts#L5

File: gtm/src/template-parameters.json [hunk 13]
```diff
@@ -0,0 +1,404 @@
...
+            "value": true
+          }
+        ]
+      },
+      {
+        "displayName": "Region",
+        "name": "region",
+        "help": "The DNS server to target (support values: de, us).",
+        "simpleValueType": true,
+        "type": "TEXT"
```
Comment (tkrugg) [inline gtm/src/template-parameters.json]: shouldn't this be a SELECT instead (only 2 values are available) ?

File: gtm/src/template-parameters.json [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+            "type": "POSITIVE_NUMBER"
+          }
+        ]
+      },
+      {
+        "displayName": "User Token",
+        "name": "setUserToken",
+        "help": "Set a persistent user identifier applied to subsequent events.",
+        "simpleValueType": true,
+        "type": "TEXT"
```
Comment (tkrugg) [inline gtm/src/template-parameters.json]: again, not sure how this can be used in this context, since it's a method, I would have expected it to appear on the same level as other methods `init`, `clickedObjectIDsAfterSearch`, etc

File: gtm/README.md [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+
+- Run the `build` command to update the [generated template](generated/search-insights.tpl)
+- Update the [changelog](CHANGELOG.md) manually
+- Commit it to GitHub
+
+For users to update the template, they need to download it again and to reimport it in the GTM interface (they won't lose their configuration).
+
+## Credits
+
+Thanks to [David Vallejo](https://www.thyngster.com/) for his initial work on the custom template.
```
Comment (redox) [inline gtm/README.md]: Thank you @thyngster !

File: gtm/src/info.json (hunk 11 above)
Comment (francoischalifour) [reply gtm/src/info.json]: This is the name of the brand, I think Algolia makes sense?

<details>
<summary>See tag types</summary>

image

</details>

Note: I'm gonna rename the template to "Search Insights" since this `displayName` already mentions Algolia.

File: gtm/src/template.js [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+    }
+
+    aa('addAlgoliaAgent', 'insights-gtm (' + TEMPLATE_VERSION + ')');
+
+    const initOptions = {
+      appId: data.appId,
+      apiKey: data.apiKey,
+      userHasOptedOut: data.userHasOptedOut,
+      region: data.region,
+      cookieDuration: data.cookieDuration,
```
[the remark this answers is not in the document]
Comment (francoischalifour) [reply gtm/src/template.js]: `undefined`

File: gtm/src/template.js [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+  case 'viewedObjectIDs': {
+    if (!isInitialized()) {
+      logger('You need to call the "init" event first.');
+      break;
+    }
+
+    const viewedObjectIDsOptions = {
+      index: data.index,
+      eventName: data.eventName,
+      objectIDs: formatValueToList(data.objectIDs),
```
[the remark this answers is not in the document]
Comment (francoischalifour) [reply gtm/src/template.js]: `objectIDs` is often a string because it can come from data attributes. We need to convert it to an array and truncate it to max 20 items (because that's the maximum amount that the API can process).

File: gtm/src/template.js (hunk 10 above)
Comment (francoischalifour) [reply gtm/src/template.js]: Sorry, there's a mistake here. It should be `data.userToken`. Good call!

I'm gonna edit the original post to add a live example.

File: gtm/src/template-parameters.json (hunk 12 above)
Comment (francoischalifour) [reply gtm/src/template-parameters.json]: Thanks!

File: gtm/src/template-parameters.json (hunk 13 above)
Comment (francoischalifour) [reply gtm/src/template-parameters.json]: Good call!

File: gtm/src/template-parameters.json (hunk 14 above)
Comment (francoischalifour) [reply gtm/src/template-parameters.json]: That's a value at this point (I'll rename it to `userToken`). If set, it calls `setUserToken`.

File: gtm/README.md (hunk 7 above)
Comment (francoischalifour) [reply gtm/README.md]: 69def5e9a9aadfa5dd433a59a70ca20a42f4f712

File: gtm/src/info.json (hunk 11 above)
Comment (tkrugg) [reply gtm/src/info.json]: ok for the original point.

About the note: on your screen shot, the display name has the brand in it for a lot of them
image
image

Comment (francoischalifour) [reply gtm/src/info.json]: That's true!

02f1f2ebba80faf873f811169253f5e0fec7e5f3
